src/GraphQL.MicrosoftDI/ScopedResolveFieldContextAdapter.cs

```diff
@@ -0,0 +1,86 @@
+using System;
+using System.Collections.Generic;
+using System.Threading;
+using GraphQL.Execution;
+using GraphQL.Instrumentation;
+using GraphQL.Language.AST;
+using GraphQL.Types;
+
+namespace GraphQL.MicrosoftDI
+{
+    internal sealed class ScopedResolveFieldContextAdapter<TSource> : IResolveFieldContext<TSource>
+    {
+        private static readonly bool _acceptNulls = !typeof(TSource).IsValueType || (typeof(TSource).IsGenericType && typeof(TSource).GetGenericTypeDefinition() == typeof(Nullable<>));
+
+        private readonly IResolveFieldContext _baseContext;
+
+        public ScopedResolveFieldContextAdapter(IResolveFieldContext baseContext, IServiceProvider serviceProvider)
+        {
+            if (baseContext.Source == null && !_acceptNulls)
+            {
+                throw new ArgumentException("baseContext.Source is null and cannot be cast to non-nullable value type " + typeof(TSource).Name, nameof(baseContext));
+            }
+            else
+            {
+                try
+                {
+                    Source = (TSource)baseContext.Source;
+                }
+                catch (InvalidCastException)
+                {
+                    throw new ArgumentException("baseContext.Source is not of type " + typeof(TSource).Name, nameof(baseContext));
+                }
+            }
+            _baseContext = baseContext ?? throw new ArgumentNullException(nameof(baseContext));
```

Comment:
This check should be the first in ctor.

Reply:
If you like.

Reply:
(The preceding code is also nothing more than a validity check also.)

Reply:
`if (baseContext.Source == null && !_acceptNulls)` gives NRE.

Reply:
oh, makes sense